opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/SystemLimitRuleConfig.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.opensearch.planner.physical;
+
+import lombok.EqualsAndHashCode;
+import lombok.RequiredArgsConstructor;
+import org.apache.calcite.adapter.enumerable.EnumerableCorrelate;
+import org.apache.calcite.adapter.enumerable.EnumerableRel;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.tools.RelBuilderFactory;
+
+@EqualsAndHashCode
+@RequiredArgsConstructor
+public class SystemLimitRuleConfig implements RelRule.Config {
+  private final int limit;
+  private RelBuilderFactory relBuilderFactory = relBuilderFactory();
+  private String description;
+  private RelRule.OperandTransform operandSupplier;
+
+  @Override
+  public RelOptRule toRule() {
+    return new OpenSearchSystemLimitRule(this, limit);
+  }
+
+  @Override
+  public RelRule.Config withRelBuilderFactory(RelBuilderFactory factory) {
+    this.relBuilderFactory = factory;
+    return this;
+  }
+
+  @Override
+  public RelRule.Config withDescription(String description) {
+    this.description = description;
+    return this;
+  }
+
+  @Override
+  public RelRule.Config withOperandSupplier(RelRule.OperandTransform transform) {
+    this.operandSupplier = transform;
+    return this;
+  }
+
+  @Override
+  public RelRule.OperandTransform operandSupplier() {
+    return this.operandSupplier;
+  }
+
+  @Override
+  public String description() {
+    return this.description;
+  }
+
+  /** Operand supplier to specify physical join operator */
+  public static RelRule.OperandTransform SYSTEM_LIMIT_JOIN_TRANSFORM =
```

Comment:
One concern with the rule-based approach is that it lacks awareness of the original command context. Consequently, it cannot manage specific commands directly. For example, an expand command translates to a correlated join internally, but since the rule doesn't recognize this translation, users cannot effectively control parameters such as system_limit.expand_size.

Is it possible apply limit when composing the RelNode? for instance Join(left, right) can be compose as Join(Limit(left, 50000), Limit(right, 50000))

Reply:
> For example, an expand command translates to a correlated join internally, but since the rule doesn't recognize this translation, users cannot effectively control parameters such as system_limit.expand_size.

The system limit rule only impacts physical plans, so whatever operators which will converted to hash join or sort merge join will be controlled. Check in-subquery in CalciteExplainIT.